src/server/pkg/hashtree/hashtree.go

```diff
@@ -0,0 +1,447 @@
+	}
+	if node.DirNode == nil {
+		return errorf(Internal, "node at \"%s\" is a file, but \"%s\" exists "+
+			"under it")
+	}
+	if !removeStr(&node.DirNode.Children, child) {
+		return errorf(Internal, "parent of \"%s\" does not contain it", path)
+	}
+	// Update hashes back to root
+	return h.visit(path, func(node *Node, parent, child string) error {
+		if node == nil {
+			return errorf(Internal,
+				"encountered orphaned file \"%s\" while deleting \"%s\"", path,
+				join(parent, child))
+		}
+		node.Size -= size
+		h.updateHash(parent)
+		return nil
+	})
+}
+
+// PutFile inserts a file into the hierarchy
+func (h *HashTree) PutFile(path string, blockRefs []*pfs.BlockRef) error {
+	h.init()
+	path = clean(path)
+
+	// Detect any path conflicts before modifying 'h'
+	if err := h.visit(path, nop); err != nil {
+		return err
+	}
+
+	// Get/Create file node to which we'll append 'blockRefs'
+	node, ok := h.Fs[path]
+	if ok {
+		if node.FileNode == nil {
+			return errorf(PathConflict, "node at \"%s\" is not a regular file ", path)
+		}
+	} else {
+		node = &Node{
+			Name:     base(path),
+			FileNode: &FileNode{},
+		}
+		h.Fs[path] = node
+	}
+
+	// Append new blocks
+	node.FileNode.BlockRefs = append(node.FileNode.BlockRefs, blockRefs...)
+	h.updateHash(path)
+
+	// Compute size growth of node (i.e. amount of data we're appending)
+	var sizeGrowth int64
+	for _, blockRef := range blockRefs {
+		sizeGrowth += int64(blockRef.Range.Upper - blockRef.Range.Lower)
+	}
+	node.Size += sizeGrowth
+
+	// Add 'path' to parent & update hashes back to root
+	return h.visit(path, func(node *Node, parent, child string) error {
+		if node == nil {
+			node = &Node{
+				Name:    base(parent),
+				Size:    0,
+				DirNode: &DirectoryNode{},
+			}
+			h.Fs[parent] = node
+		}
+		insertStr(&node.DirNode.Children, child)
+		node.Size += sizeGrowth
+		h.updateHash(parent)
+		return nil
+	})
+}
+
+// PutDir inserts an empty directory into the hierarchy
+func (h *HashTree) PutDir(path string) error {
+	h.init()
+	path = clean(path)
+
+	// Detect any path conflicts before modifying 'h'
+	if err := h.visit(path, nop); err != nil {
+		return err
+	}
+
+	// Create orphaned directory at 'path' (or end early if a directory is there)
+	if node, ok := h.Fs[path]; ok {
+		if node.DirNode == nil {
+			return errorf(PathConflict, "could not create directory at \"%s\"; a "+
+				"non-directory file is already there", path)
+		}
+		return nil
+	}
+	h.Fs[path] = &Node{
+		Name:    base(path),
+		DirNode: &DirectoryNode{},
+	}
+	h.updateHash(path)
+
+	// Add 'path' to parent & update hashes back to root
+	return h.visit(path, func(node *Node, parent, child string) error {
+		if node == nil {
+			node = &Node{
+				Name:    base(parent),
+				DirNode: &DirectoryNode{},
+			}
+			h.Fs[parent] = node
+		}
+		insertStr(&node.DirNode.Children, child)
+		h.updateHash(parent)
+		return nil
+	})
+}
+
+// DeleteFile deletes the file at 'path'.
```

Comment:
Nope, I don't think so. Simpler seems better, and they both already had the same implementation in deleteNode anyway. Fixed!